include/nccl_ofi_idpool.h

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (c) 2024 Amazon.com, Inc. or its affiliates. All rights reserved.
+ */
+
+#ifndef NCCL_OFI_IDPOOL_H_
+#define NCCL_OFI_IDPOOL_H_
+
+#ifdef _cplusplus
+extern "C" {
+#endif
+
+#include <pthread.h>
+
+#include <stdint.h>
+
+/*
+ * Pool of IDs, used to keep track of communicator IDs and MR keys.
+ */
+typedef struct nccl_ofi_idpool {
+	/* Size of the id pool (number of IDs) */
+	size_t size;
+
+	/* Number of 64-bit entries */
+	size_t entries;
+
+	/* ID pool bit array. A bit set in the array indicates
+	   that the ID corresponding to its index is available.*/
+	uint64_t *ids;
+
+	/* Lock for concurrency */
+	pthread_mutex_t lock;
+} nccl_ofi_idpool_t;
+
+/*
+ * @brief	Initialize pool of IDs
+ *
+ * Allocates and initializes a nccl_ofi_idpool_t object, marking all
+ * IDs as available.
+ *
+ * @param	idpool_p
+ *		Return value with the ID pool pointer allocated
+ * @param	size
+ *		Size of the id pool (number of IDs)
+ * @return	0 on success
+ *		non-zero on error
+ */
+int nccl_ofi_idpool_init(nccl_ofi_idpool_t *idpool, size_t size);
+
+/*
+ * @brief	Allocate an ID
+ *
+ * Extract an available ID from the ID pool, mark the ID as
+ * unavailable in the pool, and return extracted ID. No-op in case
+ * no ID was available.
+ *
+ * This operation is locked by the ID pool's internal lock.
+ *
+ * @param	idpool
+ *		The ID pool
+ * @param	id
+ *		Set to the extracted ID (zero-based), on success
+ * @return	0 on success
+ *		non-zero on error
+ */
+int nccl_ofi_idpool_allocate_id(nccl_ofi_idpool_t *idpool, uint64_t *id);
```

Comment:
What's the usecase for returning an error independent of the return?  And what's the use of a uint64_t for the id?  Just make the return value the id in an int.  Yes, that means that we can only support 2 billion ids, but that's a couple orders of magnitude more than we'll ever need.  Don't over-complicate the interface.